python-cli/common.sh

```diff
@@ -13,26 +13,45 @@ export CYAN='\033[0;36m'         # Cyan--hints and info messages
 export WHITE='\033[0;37m'        # White
 
 
-# export HARNESS_CLI_HOME if not set
-if [ -z "${HARNESS_CLI_HOME}" ]; then
-  source_path=$(readlink -f "$0" 2>/dev/null)
+## BSD/MacOS compatible readlink -f
+#
+_readlink_f() {
+  target=$1
+  cd $(dirname $target)
+  target=$(basename $target)
+
+  while [ -L "$target" ]; do
+      target=$(readlink $target)
+      cd $(dirname $target)
+      target=$(basename $target)
+  done
 
-  if [ $? -eq 0 -a -n "${source_path}" ]; then
-    # readlink has -f and suceeded
-    export HARNESS_CLI_HOME=$(dirname "${source_path%/*}/python-cli")
-  else
-    # try recursively locate while we get links
-    target_dir=$(cd `dirname "$0"`; pwd)
-    while [ -L "${target_dir}" ]; do
-      target_dir=$(ls -l "${target_dir}" | sed 's/.* -> //')
-      target_dir="$(cd `dirname "${target_dir}"`; pwd)"
-    done
-    export HARNESS_CLI_HOME="${target_dir}/python-cli"
+  physical_directory=$(pwd -P)
+  echo $physical_directory/$(basename $target)
+}
+
+#echo `basename $(dirname $(dirname $0))`
+#echo $(dirname $(dirname $0))
+#echo $(dirname $0)
+#echo $(dirname `_readlink_f $0`)
+
+# export $HARNESS_CLI_HOME if not set
+if [ -z "${HARNESS_CLI_HOME}" ]; then
+  if [ -z "$(which readlink 2>/dev/null)" ]; then
+    echo -e "${RED}readlink command must be present on your system!${NC}"
+    exit 1
   fi
+
+  #bindir=$(dirname `_readlink_f $0`)
+  # trim the last path element (which is supposed to be /bin(/))
+  #export $HARNESS_CLI_HOME=${bindir%/*}
+  export HARNESS_CLI_HOME=$(dirname `_readlink_f $0`)
```

Comment:
@dennybaa 

I have a symlink:
```sudo ln -s /Users/pat/harness-cli/python-cli /usr/local/harness-cli```
The symlink is first in my PATH before the embedded harness cli

This works on macOS, not sure why you are trimming the last subdir off the env var. In this case HARNESS_CLI_HOME would be `/Users/pat/harness-cli` so every reference to a file to be sourced or executed would have to be prefixed with `/python-cli` to make `${HARNESS_HOME}/python-cli/common.sh` and this clearly makes no sense so I removed the trimming for this CLI but now I wonder if this does not work on Linux.

I also do not understand why the code DOES work for the Makelfile build of harness on macOS and linux. I bet it is broken now for the script build of harness with `make-harness-distribution.sh`. 

The env var `HARNESS_HOME` or `HARNESS_CLI_HOME` should be the full physical filesystem path (no symlink) to the `bin` directory of harness or the `python-cli` directory of this harness-cli -- right? BTW I still use your `_readlink_f` but we must have different ideas for where project HOME points.

What do you think the value of `HARNESS_HOME` should be?

Reply:
Maybe our problem is that common.sh is sourced twice and calculates completely different values for HARNESS_CLI_HOME here and also for the embedded cli

Reply:
The initial script aimed to locate the `HARNESS_HOME` which is not required in this case. Exporting `HARNESS_CLI_HOME` seems should not happen at all. Splitting functions and common into two files is an overkill too.